File: src/main/java/org/taktik/icure/logic/impl/GroupLogicImpl.java
package org.taktik.icure.logic.impl;

import org.ektorp.CouchDbConnector;
import org.ektorp.CouchDbInstance;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.core.task.TaskExecutor;
import org.springframework.stereotype.Service;
import org.taktik.icure.dao.GroupDAO;
import org.taktik.icure.entities.Group;
import org.taktik.icure.entities.Replication;
import org.taktik.icure.entities.base.Security;
import org.taktik.icure.entities.base.User;
import org.taktik.icure.entities.embed.DatabaseSynchronization;
import org.taktik.icure.logic.ReplicationLogic;
import org.taktik.icure.logic.SessionLogic;
import org.taktik.icure.logic.UserLogic;
import org.taktik.icure.properties.CouchDbProperties;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class GroupLogicImpl implements org.taktik.icure.logic.GroupLogic {
	public static final String ADMIN_GROUP = "xx-f203c688-159c-4d7e-8453-c5f24dce8418";
	private SessionLogic sessionLogic;
	private GroupDAO groupDAO;
	private CouchDbInstance couchdbInstance;
	private UserLogic userLogic;
	private ReplicationLogic replicationLogic;
	private CouchDbProperties couchDbProperties;
	private TaskExecutor threadPoolTaskExecutor;

	@Autowired
	public void setCouchDbProperties(CouchDbProperties couchDbProperties) {
		this.couchDbProperties = couchDbProperties;
	}

	@Autowired
	public void setReplicationLogic(ReplicationLogic replicationLogic) {
		this.replicationLogic = replicationLogic;
	}

	@Autowired
	public void setUserLogic(UserLogic userLogic) {
		this.userLogic = userLogic;
	}

	@Autowired
	public void setSessionLogic(SessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}

	@Autowired
	public void setGroupDAO(GroupDAO groupDAO) {
		this.groupDAO = groupDAO;
	}

	@Autowired
	public void setCouchdbInstance(CouchDbInstance couchdbInstance) {
		this.couchdbInstance = couchdbInstance;
	}

	@Autowired
	public void setThreadPoolTaskExecutor(@Qualifier("threadPoolTaskExecutor") TaskExecutor threadPoolTaskExecutor) {
		this.threadPoolTaskExecutor = threadPoolTaskExecutor;
	}

	@Override
	public Group createGroup(Group group, Replication initialReplication) throws IllegalAccessException {
		String id = sessionLogic.getCurrentSessionContext().getGroupIdUserId();
		if (id == null) {
			throw new IllegalAccessException("No registered user");
		}
		if (!ADMIN_GROUP.equals(userLogic.getUserOnFallbackDb(id).getGroupId())) {
			throw new IllegalAccessException("No registered user");
		}

		List<String> paths = Arrays.asList(
				"icure-" + group.getId() + "-base",
				"icure-" + group.getId() + "-patient",
				"icure-" + group.getId() + "-healthdata"
		);

		List<DatabaseSynchronization> sanitizedDatabaseSynchronizations = initialReplication != null ? initialReplication.getDatabaseSynchronizations().stream().filter(ds -> {
			try {
				URI couch = new URI(couchDbProperties.getUrl());
				URI dest = new URI(ds.getTarget());

				if (!(dest.getPort() == 443 || dest.getPort() == 5984 || dest.getPort() == -1 && dest.getScheme().equals("https"))) {
					throw new IllegalArgumentException("Cannot start replication: invalid destination port (must be 5984 or 443)");
				}
				if (!(dest.getHost().equals(couch.getHost()) || dest.getHost().equals("127.0.0.1") || dest.getHost().equals("localhost"))) {
					throw new IllegalArgumentException("Cannot start replication: invalid destination " + dest.getHost() + "(must be " + couch.getHost() + " or localhost/127.0.0.1 )");
				}
				if (paths.stream().noneMatch(p -> dest.getPath().startsWith("/" + p))) {
					throw new IllegalArgumentException("Cannot start replication: invalid destination path " + dest.getPath() + " ( must match start with any of /" + String.join(", /", paths));
				}
				return true;
			} catch (URISyntaxException e) {
				throw new IllegalArgumentException("Cannot start replication: invalid target");
			}
		}).collect(Collectors.toList()) : null;

		User dbUser = new User(group.getId(), group.getPassword());
		couchdbInstance.createConnector("_users", false).create("org.couchdb.user:" + group.getId(), dbUser);

		Security security = new Security(group.getId());

		paths.forEach(c -> {
			CouchDbConnector connector = couchdbInstance.createConnector(c, true);
			connector.create("_security", security);
		});

		Group result = groupDAO.save(group);

		if (initialReplication != null) {
			initialReplication.setDatabaseSynchronizations(sanitizedDatabaseSynchronizations);
			threadPoolTaskExecutor.execute(() -> replicationLogic.startDatabaseSynchronisations(initialReplication, false));
		}

		return result.getRev() != null ? result : null;
	}

	@Override
	public Group findGroup(String groupId) {
		return groupDAO.find(groupId);
	}

}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/GroupLogicImpl.kt
package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.flow
import org.eclipse.jetty.client.HttpClient
import org.springframework.core.task.TaskExecutor
import org.springframework.stereotype.Service
import org.taktik.couchdb.ClientImpl
import org.taktik.couchdb.create
import org.taktik.icure.asyncdao.GroupDAO
import org.taktik.icure.asyncdao.UserDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.GroupLogic
import org.taktik.icure.entities.Group
import org.taktik.icure.entities.Replication
import org.taktik.icure.entities.User
import org.taktik.icure.entities.base.Security
import org.taktik.icure.entities.embed.DatabaseSynchronization
import org.taktik.icure.properties.CouchDbProperties
import java.net.URI
import java.net.URISyntaxException

@Service
class GroupLogicImpl(private val httpClient: HttpClient,
                     private val sessionLogic: AsyncSessionLogic,
                     private val groupDAO: GroupDAO,
                     private val userDAO: UserDAO,
                     private val couchDbProperties: CouchDbProperties,
                     private val threadPoolTaskExecutor: TaskExecutor) : GroupLogic {

    private val dbInstanceUri = URI(couchDbProperties.url)

    override suspend fun createGroup(
            id: String,
            name: String,
            password: String,
            server: String?,
            q: Int?,
            n: Int?,
            initialReplication: Replication?
    ): Group? {
        val groupUserId = sessionLogic.getCurrentSessionContext().getGroupIdUserId()
        val groupId = userDAO.getOnFallback(dbInstanceUri, groupUserId, false)?.groupId ?: throw IllegalAccessException("Invalid user, no group")
        val userGroup = this.groupDAO.get(groupId)
        if (userGroup == null || (groupId != ADMIN_GROUP && !userGroup.isSuperAdmin)) {
            throw IllegalAccessException("No registered super admin user")
        }
        val paths = listOf(
                "icure-$id-base",
                "icure-$id-patient",
                "icure-$id-healthdata"
        )
        val sanitizedDatabaseSynchronizations = initialReplication?.databaseSynchronizations?.filter { ds: DatabaseSynchronization ->
            try {
                val couch = URI(couchDbProperties.url)
                val dest = URI(ds.target)
                require(dest.port == 443 || dest.port == 5984 || dest.port == -1 && dest.scheme == "https") { "Cannot start replication: invalid destination port (must be 5984 or 443)" }
                require(dest.host == couch.host || dest.host == "127.0.0.1" || dest.host == "localhost") { "Cannot start replication: invalid destination " + dest.host + "(must be " + couch.host + " or localhost/127.0.0.1 )" }
                require(!paths.stream().noneMatch { p: String -> dest.path.startsWith("/$p") }) { "Cannot start replication: invalid destination path " + dest.path + " ( must match start with any of /" + java.lang.String.join(", /", paths) }
                true
            } catch (e: URISyntaxException) {
                throw IllegalArgumentException("Cannot start replication: invalid target")
            }
        } ?: listOf()
        val dbUser = User("org.couchdb.user:$id", id, password)
        val security = Security(id)
        val group = Group(id, name, password).apply {
            superGroup = groupId
            superAdmin = groupId == ADMIN_GROUP
            server?.let { sv -> servers = couchDbProperties.altUrlsList().filter { it.contains(sv) } }
        }

        val servers = if (group.servers?.isNotEmpty() == true) group.servers else listOf(couchDbProperties.url)
        servers.forEach {
            ClientImpl(httpClient, org.ektorp.http.URI.of(it).append("_users"), couchDbProperties.username!!, couchDbProperties.password!!).create(dbUser)
            paths.forEach { c ->
                val client = ClientImpl(httpClient, org.ektorp.http.URI.of(it).append(c), couchDbProperties.username!!, couchDbProperties.password!!)
                if (client.create(if (c.endsWith("-base")) 1 else q, n)) {
                    client.security(security)
                }
            }
        }

        val result = groupDAO.save(group)
        return if (result?.rev != null) result else null
    }

    override suspend fun getGroup(groupId: String): Group? {
        return groupDAO.get(groupId)
    }

    override fun listGroups(): Flow<Group> = flow {
        val groupUserId = sessionLogic.getCurrentSessionContext().getGroupIdUserId()
        val groupId = userDAO.getOnFallback(dbInstanceUri, groupUserId, false)?.groupId ?: throw IllegalAccessException("Invalid user, no group")

        emitAll(groupDAO.getAll().filter { it.superGroup == groupId  })
    }

    companion object {
        const val ADMIN_GROUP = "xx-f203c688-159c-4d7e-8453-c5f24dce8418"
    }
}
